home/vasco-santos/2021-03-exploration-report-1-on-js-lotus-client.md

```diff
@@ -0,0 +1,102 @@
+# Experiment with lotus client
+
+## Deal Flow
+
+Requisites:
+- A wallet id with FIL
+
+Flow:
+1. import file
+  * `provider.importFile(data: String): Promise<CID>`
+2. fetch file info for deal
+  * `client.clientDealSize(cid: Cid): Promise<DataSize>`
+3. list miners
+  * `client.stateListMiners(tipSetKey: Cid[]): Promise<Array<string>>`
+4. get miner info for each miner
+  * `client.stateMinerInfo(address: string, tipSetKey: Cid[]): Promise<MinerInfo>`
+5. find the price and conditions of each miner
+  * `client.clientQueryAsk(id: string, address: string): Promise<StorageAsk>`
+6. start a deal with the intended miners
+  * `clientStartDeal (startDealParams: StartDealParams): Promise<Cid>`
+
+
+```js
+// const wallet = ...
+const cidString = await provider.importFile('this is random text!')
+
+const cid = {
+  '/': cidString
+}
+
+const dataSize = await client.clientDealSize(cid)
+
+const head = await client.chainHead()
+const miners = await client.stateListMiners(head.Cids)
+
+for (const miner in miners) {
+  let minerInfo, storageAsk
+  try {
+    minerInfo = await client.stateMinerInfo(miner, head.Cids)
```

Comment:
Any idea what would happen with the RPC & the library if we did `Promise.all()` for these queries and ran them in parallel? That's what Lotus does internally when you `list-asks`.

Reply:
I did not try it out, I don't think there are any limits, we should have a look. Honestly this was mostly pseudocode, in my local experiences I selected a miner manually and from there did not iterate all the miners because it was super slow

Reply:
a problem that sentinel hit was that asking lotus for lots of data via the api makes it fall behind on syncing the chain. This is likely only a issue at sentinel scale where you are asking the node about **everything** that happened at an epoch, every epoch, but worth being aware of.

Reply:
that's an interesting enough anecdote to add to our report, have added it [here](https://github.com/protocol/w3dt-sudo/issues/9)
I don't think we want to delve too deeply into this, but I wonder if the lotus-gateway does anything to help protect against such problems?